staging/src/k8s.io/apiserver/pkg/util/proxy/websocket.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/httpstream/wsstream"
+	constants "k8s.io/apimachinery/pkg/util/remotecommand"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+const (
+	// idleTimeout is the read/write deadline set for websocket server connection. Reading
+	// or writing the connection will return an i/o timeout if this deadline is exceeded.
+	// Currently, we use the same value as the kubelet websocket server.
+	defaultIdleConnectionTimeout = 4 * time.Hour
+
+	// Deadline for writing errors to the websocket connection before io/timeout.
+	writeErrorDeadline = 10 * time.Second
+)
+
+// Options contains details about which streams are required for
+// remote command execution.
+type Options struct {
+	Stdin  bool
+	Stdout bool
+	Stderr bool
+	Tty    bool
+}
+
+// conns contains the connection and streams used when
+// forwarding an attach or execute session into a container.
+type conns struct {
+	conn         io.Closer
+	stdinStream  io.ReadCloser
+	stdoutStream io.WriteCloser
+	stderrStream io.WriteCloser
+	writeStatus  func(status *apierrors.StatusError) error
+	resizeStream io.ReadCloser
+	resizeChan   chan remotecommand.TerminalSize
+	tty          bool
+}
+
+// Create WebSocket server streams to respond to a WebSocket client. Creates the streams passed
+// in the stream options.
+func webSocketServerStreams(req *http.Request, w http.ResponseWriter, opts Options) (*conns, error) {
+	ctx, err := createWebSocketStreams(req, w, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if ctx.resizeStream != nil {
+		ctx.resizeChan = make(chan remotecommand.TerminalSize)
+		go handleResizeEvents(req.Context(), ctx.resizeStream, ctx.resizeChan)
+	}
+
+	return ctx, nil
+}
+
+// Read terminal resize events off of passed stream and queue into passed channel.
+func handleResizeEvents(ctx context.Context, stream io.Reader, channel chan<- remotecommand.TerminalSize) {
+	defer runtime.HandleCrash()
```

Comment:
(this is true for all goroutines spawned by things the translating stream handler is using, websocket or spdy)

Reply:
1. Panic propagation already exists for most of the goroutines spawned from the stream translator proxy. A panic propagation channel is used in SPDY executor `StreamWithContext()` for this purpose.
2. I have added another panic propagation channel for the goroutine which is spawned for the websocket server in `conn.go`.
3. For this `handleResizeEvents()`, I have added a deferred panic/recovery function to ensure this spawned goroutine can not crash the server. This function only logs an error now, because it does not appear possible to wait for a panic channel in this instance.

Please have a look and let me know what you think.